pyvo/dal/tests/test_datalink.py

```diff
@@ -128,6 +128,7 @@ def _debytify(v):
 @pytest.mark.filterwarnings("ignore::astropy.io.votable.exceptions.W06")
 @pytest.mark.filterwarnings("ignore::astropy.io.votable.exceptions.W48")
 @pytest.mark.filterwarnings("ignore::astropy.io.votable.exceptions.E02")
+@pytest.mark.remote_data
```

Comment:
These are not remote data tests. They are mocked and return content of file in `pyvo/dal/tests/data`

Reply:
hmm, interesting, I can't recall why I marked them as remote, maybe the mocking wasn't working properly? I'll check on this.

Reply:
yeap, indeed:

```
======================================================= FAILURES =======================================================
____________________________________ TestSemanticsRetrieval.test_access_with_string ____________________________________

self = <urllib.request.HTTPHandler object at 0x132ad4280>, http_class = <class 'http.client.HTTPConnection'>
req = <urllib.request.Request object at 0x132ad41c0>, http_conn_args = {}, host = 'www.ivoa.net'
h = <http.client.HTTPConnection object at 0x132ad46a0>

    def do_open(self, http_class, req, **http_conn_args):
        """Return an HTTPResponse object for the request, using http_class.
    
        http_class must implement the HTTPConnection API from http.client.
        """
        host = req.host
        if not host:
            raise URLError('no host given')
    
        # will parse host:port
        h = http_class(host, timeout=req.timeout, **http_conn_args)
        h.set_debuglevel(self._debuglevel)
    
        headers = dict(req.unredirected_hdrs)
        headers.update({k: v for k, v in req.headers.items()
                        if k not in headers})
    
        # TODO(jhylton): Should this be redesigned to handle
        # persistent connections?
    
        # We want to make an HTTP/1.1 request, but the addinfourl
        # class isn't prepared to deal with a persistent connection.
        # It will try to read all remaining data from the socket,
        # which will block while the server waits for the next request.
        # So make sure the connection gets closed after the (only)
        # request.
        headers["Connection"] = "close"
        headers = {name.title(): val for name, val in headers.items()}
    
        if req._tunnel_host:
            tunnel_headers = {}
            proxy_auth_hdr = "Proxy-Authorization"
            if proxy_auth_hdr in headers:
                tunnel_headers[proxy_auth_hdr] = headers[proxy_auth_hdr]
                # Proxy-Authorization should not be sent to origin
                # server.
                del headers[proxy_auth_hdr]
            h.set_tunnel(req._tunnel_host, headers=tunnel_headers)
    
        try:
            try:
>               h.request(req.get_method(), req.selector, req.data, headers,
                          encode_chunked=req.has_header('Transfer-encoding'))

../../../.pyenv/versions/3.9.1/lib/python3.9/urllib/request.py:1342: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <http.client.HTTPConnection object at 0x132ad46a0>, method = 'GET', url = '/rdf/datalink/core', body = None
headers = {'Accept': 'application/x-desise+json', 'Connection': 'close', 'Host': 'www.ivoa.net', 'User-Agent': 'Python-urllib/3.9'}

    def request(self, method, url, body=None, headers={}, *,
                encode_chunked=False):
        """Send a complete request to the server."""
>       self._send_request(method, url, body, headers, encode_chunked)

../../../.pyenv/versions/3.9.1/lib/python3.9/http/client.py:1255: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <http.client.HTTPConnection object at 0x132ad46a0>, method = 'GET', url = '/rdf/datalink/core', body = None
headers = {'Accept': 'application/x-desise+json', 'Connection': 'close', 'Host': 'www.ivoa.net', 'User-Agent': 'Python-urllib/3.9'}
encode_chunked = False

    def _send_request(self, method, url, body, headers, encode_chunked):
        # Honor explicitly requested Host: and Accept-Encoding: headers.
        header_names = frozenset(k.lower() for k in headers)
        skips = {}
        if 'host' in header_names:
            skips['skip_host'] = 1
        if 'accept-encoding' in header_names:
            skips['skip_accept_encoding'] = 1
    
        self.putrequest(method, url, **skips)
    
        # chunked encoding will happen if HTTP/1.1 is used and either
        # the caller passes encode_chunked=True or the following
        # conditions hold:
        # 1. content-length has not been explicitly set
        # 2. the body is a file or iterable, but not a str or bytes-like
        # 3. Transfer-Encoding has NOT been explicitly set by the caller
    
        if 'content-length' not in header_names:
            # only chunk body if not explicitly set for backwards
            # compatibility, assuming the client code is already handling the
            # chunking
            if 'transfer-encoding' not in header_names:
                # if content-length cannot be automatically determined, fall
                # back to chunked encoding
                encode_chunked = False
                content_length = self._get_content_length(body, method)
                if content_length is None:
                    if body is not None:
                        if self.debuglevel > 0:
                            print('Unable to determine size of %r' % body)
                        encode_chunked = True
                        self.putheader('Transfer-Encoding', 'chunked')
                else:
                    self.putheader('Content-Length', str(content_length))
        else:
            encode_chunked = False
    
        for hdr, value in headers.items():
            self.putheader(hdr, value)
        if isinstance(body, str):
            # RFC 2616 Section 3.7.1 says that text default has a
            # default charset of iso-8859-1.
            body = _encode(body, 'body')
>       self.endheaders(body, encode_chunked=encode_chunked)

../../../.pyenv/versions/3.9.1/lib/python3.9/http/client.py:1301: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <http.client.HTTPConnection object at 0x132ad46a0>, message_body = None

    def endheaders(self, message_body=None, *, encode_chunked=False):
        """Indicate that the last header line has been sent to the server.
    
        This method sends the request to the server.  The optional message_body
        argument can be used to pass a message body associated with the
        request.
        """
        if self.__state == _CS_REQ_STARTED:
            self.__state = _CS_REQ_SENT
        else:
            raise CannotSendHeader()
>       self._send_output(message_body, encode_chunked=encode_chunked)

../../../.pyenv/versions/3.9.1/lib/python3.9/http/client.py:1250: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <http.client.HTTPConnection object at 0x132ad46a0>, message_body = None, encode_chunked = False

    def _send_output(self, message_body=None, encode_chunked=False):
        """Send the currently buffered request and clear the buffer.
    
        Appends an extra \\r\\n to the buffer.
        A message_body may be specified, to be appended to the request.
        """
        self._buffer.extend((b"", b""))
        msg = b"\r\n".join(self._buffer)
        del self._buffer[:]
>       self.send(msg)

../../../.pyenv/versions/3.9.1/lib/python3.9/http/client.py:1010: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <http.client.HTTPConnection object at 0x132ad46a0>
data = b'GET /rdf/datalink/core HTTP/1.1\r\nAccept-Encoding: identity\r\nHost: www.ivoa.net\r\nUser-Agent: Python-urllib/3.9\r\nAccept: application/x-desise+json\r\nConnection: close\r\n\r\n'

    def send(self, data):
        """Send `data' to the server.
        ``data`` can be a string object, a bytes object, an array object, a
        file-like object that supports a .read() method, or an iterable object.
        """
    
        if self.sock is None:
            if self.auto_open:
>               self.connect()

../../../.pyenv/versions/3.9.1/lib/python3.9/http/client.py:950: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <http.client.HTTPConnection object at 0x132ad46a0>

    def connect(self):
        """Connect to the host and port specified in __init__."""
>       self.sock = self._create_connection(
            (self.host,self.port), self.timeout, self.source_address)

../../../.pyenv/versions/3.9.1/lib/python3.9/http/client.py:921: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = (('www.ivoa.net', 80), 10.0, None), kwargs = {}, host = 'www.ivoa.net', addr_arg = 0
valid_hosts = {'127.0.0.1', 'localhost'}, hostname = 'kecske3.local'
fqdn = '1.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.ip6.arpa'
host_ips = {'140.105.76.131', 'www.ivoa.net'}

    def new_function(*args, **kwargs):
        if isinstance(args[0], socket.socket):
            if not args[0].family in (socket.AF_INET, socket.AF_INET6):
                # Should be fine in all but some very obscure cases
                # More to the point, we don't want to affect AF_UNIX
                # sockets.
                return original_function(*args, **kwargs)
            host = args[1][0]
            addr_arg = 1
            valid_hosts = set(['localhost', '127.0.0.1', '::1'])
        else:
            # The only other function this is used to wrap currently is
            # socket.create_connection, which should be passed a 2-tuple, but
            # we'll check just in case
            if not (isinstance(args[0], tuple) and len(args[0]) == 2):
                return original_function(*args, **kwargs)
    
            host = args[0][0]
            addr_arg = 0
            valid_hosts = set(['localhost', '127.0.0.1'])
    
        # Astropy + GitHub data
        if allow_astropy_data:
            for valid_host in ASTROPY_HOSTS:
                valid_hosts = valid_hosts.union(_resolve_host_ips(valid_host))
    
        # Only GitHub data
        elif allow_github_data:
            for valid_host in GITHUB_HOSTS:
                valid_hosts = valid_hosts.union(_resolve_host_ips(valid_host))
    
        hostname = socket.gethostname()
        fqdn = socket.getfqdn()
    
        if host in (hostname, fqdn):
            host = 'localhost'
            host_ips = set([host])
            new_addr = (host, args[addr_arg][1])
            args = args[:addr_arg] + (new_addr,) + args[addr_arg + 1:]
        else:
            host_ips = _resolve_host_ips(host)
    
        if len(host_ips & valid_hosts) > 0:  # Any overlap is acceptable
            return original_function(*args, **kwargs)
        else:
>           raise IOError("An attempt was made to connect to the internet "
                          "by a test that was not marked `remote_data`. The "
                          "requested host was: {0}".format(host))
E           OSError: An attempt was made to connect to the internet by a test that was not marked `remote_data`. The requested host was: www.ivoa.net

../../../.pyenv/versions/3.9.1/lib/python3.9/site-packages/pytest_remotedata/disable_internet.py:99: OSError

During handling of the above exception, another exception occurred:

voc_name = 'datalink/core', force_update = False

    @functools.lru_cache()
    def get_vocabulary(voc_name, force_update=False):
        """returns an IVOA vocabulary in its "desise" form.
    
        See Vocabularies in the VO 2 to see what is inside of this.
    
        This will use a cache to avoid repeated updates, but it
        will attempt to re-download if the cached copy is older than 6 months.
        """
        src_url = IVOA_VOCABULARY_ROOT+voc_name
        if force_update:
            clear_download_cache(src_url)
    
        try:
>           src_name = download_file(
                src_url,
                cache=True,
                show_progress=False,
                http_headers={"accept": "application/x-desise+json"})

pyvo/utils/vocabularies.py:40: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

remote_url = 'http://www.ivoa.net/rdf/datalink/core', cache = True, show_progress = False, timeout = 10.0
sources = ['http://www.ivoa.net/rdf/datalink/core'], pkgname = 'astropy'
http_headers = {'accept': 'application/x-desise+json'}, ssl_context = None, allow_insecure = False

    def download_file(remote_url, cache=False, show_progress=True, timeout=None,
                      sources=None, pkgname='astropy', http_headers=None,
                      ssl_context=None, allow_insecure=False):
        """Downloads a URL and optionally caches the result.
    
        It returns the filename of a file containing the URL's contents.
        If ``cache=True`` and the file is present in the cache, just
        returns the filename; if the file had to be downloaded, add it
        to the cache. If ``cache="update"`` always download and add it
        to the cache.
    
        The cache is effectively a dictionary mapping URLs to files; by default the
        file contains the contents of the URL that is its key, but in practice
        these can be obtained from a mirror (using ``sources``) or imported from
        the local filesystem (using `~import_file_to_cache` or
        `~import_download_cache`).  Regardless, each file is regarded as
        representing the contents of a particular URL, and this URL should be used
        to look them up or otherwise manipulate them.
    
        The files in the cache directory are named according to a cryptographic
        hash of their URLs (currently MD5, so hackers can cause collisions).
        The modification times on these files normally indicate when they were
        last downloaded from the Internet.
    
        Parameters
        ----------
        remote_url : str
            The URL of the file to download
    
        cache : bool or "update", optional
            Whether to cache the contents of remote URLs. If "update",
            always download the remote URL in case there is a new version
            and store the result in the cache.
    
        show_progress : bool, optional
            Whether to display a progress bar during the download (default
            is `True`). Regardless of this setting, the progress bar is only
            displayed when outputting to a terminal.
    
        timeout : float, optional
            Timeout for remote requests in seconds (default is the configurable
            `astropy.utils.data.Conf.remote_timeout`).
    
        sources : list of str, optional
            If provided, a list of URLs to try to obtain the file from. The
            result will be stored under the original URL. The original URL
            will *not* be tried unless it is in this list; this is to prevent
            long waits for a primary server that is known to be inaccessible
            at the moment. If an empty list is passed, then ``download_file``
            will not attempt to connect to the Internet, that is, if the file
            is not in the cache a KeyError will be raised.
    
        pkgname : `str`, optional
            The package name to use to locate the download cache. i.e. for
            ``pkgname='astropy'`` the default cache location is
            ``~/.astropy/cache``.
    
        http_headers : dict or None
            HTTP request headers to pass into ``urlopen`` if needed. (These headers
            are ignored if the protocol for the ``name_or_obj``/``sources`` entry
            is not a remote HTTP URL.) In the default case (None), the headers are
            ``User-Agent: some_value`` and ``Accept: */*``, where ``some_value``
            is set by ``astropy.utils.data.conf.default_http_user_agent``.
    
        ssl_context : dict, optional
            Keyword arguments to pass to `ssl.create_default_context` when
            downloading from HTTPS or TLS+FTP sources.  This can be used provide
            alternative paths to root CA certificates.  Additionally, if the key
            ``'certfile'`` and optionally ``'keyfile'`` and ``'password'`` are
            included, they are passed to `ssl.SSLContext.load_cert_chain`.  This
            can be used for performing SSL/TLS client certificate authentication
            for servers that require it.
    
        allow_insecure : bool, optional
            Allow downloading files over a TLS/SSL connection even when the server
            certificate verification failed.  When set to `True` the potentially
            insecure download is allowed to proceed, but an
            `~astropy.utils.exceptions.AstropyWarning` is issued.  If you are
            frequently getting certificate verification warnings, consider
            installing or upgrading `certifi`_ package, which provides frequently
            updated certificates for common root CAs (i.e., a set similar to those
            used by web browsers).  If installed, Astropy will use it
            automatically.
    
            .. _certifi: https://pypi.org/project/certifi/
    
        Returns
        -------
        local_path : str
            Returns the local path that the file was download to.
    
        Raises
        ------
        urllib.error.URLError
            Whenever there's a problem getting the remote file.
        KeyError
            When a file was requested from the cache but is missing and no
            sources were provided to obtain it from the Internet.
    
        Notes
        -----
        Because this function returns a filename, another process could run
        `clear_download_cache` before you actually open the file, leaving
        you with a filename that no longer points to a usable file.
        """
        if timeout is None:
            timeout = conf.remote_timeout
        if sources is None:
            sources = [remote_url]
        if http_headers is None:
            http_headers = {'User-Agent': conf.default_http_user_agent,
                            'Accept': '*/*'}
    
        missing_cache = ""
    
        url_key = remote_url
    
        if cache:
            try:
                dldir = _get_download_cache_loc(pkgname)
            except OSError as e:
                cache = False
                missing_cache = (
                    f"Cache directory cannot be read or created ({e}), "
                    f"providing data in temporary file instead."
                )
            else:
                if cache == "update":
                    pass
                elif isinstance(cache, str):
                    raise ValueError(f"Cache value '{cache}' was requested but "
                                     f"'update' is the only recognized string; "
                                     f"otherwise use a boolean")
                else:
                    filename = os.path.join(dldir, _url_to_dirname(url_key), "contents")
                    if os.path.exists(filename):
                        return os.path.abspath(filename)
    
        errors = {}
        for source_url in sources:
            try:
                f_name = _download_file_from_source(
                        source_url,
                        timeout=timeout,
                        show_progress=show_progress,
                        cache=cache,
                        remote_url=remote_url,
                        pkgname=pkgname,
                        http_headers=http_headers,
                        ssl_context=ssl_context,
                        allow_insecure=allow_insecure)
                # Success!
                break
    
            except urllib.error.URLError as e:
                # errno 8 is from SSL "EOF occurred in violation of protocol"
                if (hasattr(e, 'reason')
                        and hasattr(e.reason, 'errno')
                        and e.reason.errno == 8):
                    e.reason.strerror = (e.reason.strerror +
                                         '. requested URL: '
                                         + remote_url)
                    e.reason.args = (e.reason.errno, e.reason.strerror)
                errors[source_url] = e
            except socket.timeout as e:
                # this isn't supposed to happen, but occasionally a socket.timeout
                # gets through.  It's supposed to be caught in urllib and raised
                # in this way, but for some reason in mysterious circumstances it
                # doesn't (or didn't in python2?). So we'll just re-raise it here
                # instead.
                errors[source_url] = e
        else:   # No success
            if not sources:
                raise KeyError(
                    f"No sources listed and file {remote_url} not in cache! "
                    f"Please include primary URL in sources if you want it to be "
                    f"included as a valid source.")
            elif len(sources) == 1:
>               raise errors[sources[0]]

../../../.pyenv/versions/3.9.1/lib/python3.9/site-packages/astropy/utils/data.py:1396: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

remote_url = 'http://www.ivoa.net/rdf/datalink/core', cache = True, show_progress = False, timeout = 10.0
sources = ['http://www.ivoa.net/rdf/datalink/core'], pkgname = 'astropy'
http_headers = {'accept': 'application/x-desise+json'}, ssl_context = None, allow_insecure = False

    def download_file(remote_url, cache=False, show_progress=True, timeout=None,
                      sources=None, pkgname='astropy', http_headers=None,
                      ssl_context=None, allow_insecure=False):
        """Downloads a URL and optionally caches the result.
    
        It returns the filename of a file containing the URL's contents.
        If ``cache=True`` and the file is present in the cache, just
        returns the filename; if the file had to be downloaded, add it
        to the cache. If ``cache="update"`` always download and add it
        to the cache.
    
        The cache is effectively a dictionary mapping URLs to files; by default the
        file contains the contents of the URL that is its key, but in practice
        these can be obtained from a mirror (using ``sources``) or imported from
        the local filesystem (using `~import_file_to_cache` or
        `~import_download_cache`).  Regardless, each file is regarded as
        representing the contents of a particular URL, and this URL should be used
        to look them up or otherwise manipulate them.
    
        The files in the cache directory are named according to a cryptographic
        hash of their URLs (currently MD5, so hackers can cause collisions).
        The modification times on these files normally indicate when they were
        last downloaded from the Internet.
    
        Parameters
        ----------
        remote_url : str
            The URL of the file to download
    
        cache : bool or "update", optional
            Whether to cache the contents of remote URLs. If "update",
            always download the remote URL in case there is a new version
            and store the result in the cache.
    
        show_progress : bool, optional
            Whether to display a progress bar during the download (default
            is `True`). Regardless of this setting, the progress bar is only
            displayed when outputting to a terminal.
    
        timeout : float, optional
            Timeout for remote requests in seconds (default is the configurable
            `astropy.utils.data.Conf.remote_timeout`).
    
        sources : list of str, optional
            If provided, a list of URLs to try to obtain the file from. The
            result will be stored under the original URL. The original URL
            will *not* be tried unless it is in this list; this is to prevent
            long waits for a primary server that is known to be inaccessible
            at the moment. If an empty list is passed, then ``download_file``
            will not attempt to connect to the Internet, that is, if the file
            is not in the cache a KeyError will be raised.
    
        pkgname : `str`, optional
            The package name to use to locate the download cache. i.e. for
            ``pkgname='astropy'`` the default cache location is
            ``~/.astropy/cache``.
    
        http_headers : dict or None
            HTTP request headers to pass into ``urlopen`` if needed. (These headers
            are ignored if the protocol for the ``name_or_obj``/``sources`` entry
            is not a remote HTTP URL.) In the default case (None), the headers are
            ``User-Agent: some_value`` and ``Accept: */*``, where ``some_value``
            is set by ``astropy.utils.data.conf.default_http_user_agent``.
    
        ssl_context : dict, optional
            Keyword arguments to pass to `ssl.create_default_context` when
            downloading from HTTPS or TLS+FTP sources.  This can be used provide
            alternative paths to root CA certificates.  Additionally, if the key
            ``'certfile'`` and optionally ``'keyfile'`` and ``'password'`` are
            included, they are passed to `ssl.SSLContext.load_cert_chain`.  This
            can be used for performing SSL/TLS client certificate authentication
            for servers that require it.
    
        allow_insecure : bool, optional
            Allow downloading files over a TLS/SSL connection even when the server
            certificate verification failed.  When set to `True` the potentially
            insecure download is allowed to proceed, but an
            `~astropy.utils.exceptions.AstropyWarning` is issued.  If you are
            frequently getting certificate verification warnings, consider
            installing or upgrading `certifi`_ package, which provides frequently
            updated certificates for common root CAs (i.e., a set similar to those
            used by web browsers).  If installed, Astropy will use it
            automatically.
    
            .. _certifi: https://pypi.org/project/certifi/
    
        Returns
        -------
        local_path : str
            Returns the local path that the file was download to.
    
        Raises
        ------
        urllib.error.URLError
            Whenever there's a problem getting the remote file.
        KeyError
            When a file was requested from the cache but is missing and no
            sources were provided to obtain it from the Internet.
    
        Notes
        -----
        Because this function returns a filename, another process could run
        `clear_download_cache` before you actually open the file, leaving
        you with a filename that no longer points to a usable file.
        """
        if timeout is None:
            timeout = conf.remote_timeout
        if sources is None:
            sources = [remote_url]
        if http_headers is None:
            http_headers = {'User-Agent': conf.default_http_user_agent,
                            'Accept': '*/*'}
    
        missing_cache = ""
    
        url_key = remote_url
    
        if cache:
            try:
                dldir = _get_download_cache_loc(pkgname)
            except OSError as e:
                cache = False
                missing_cache = (
                    f"Cache directory cannot be read or created ({e}), "
                    f"providing data in temporary file instead."
                )
            else:
                if cache == "update":
                    pass
                elif isinstance(cache, str):
                    raise ValueError(f"Cache value '{cache}' was requested but "
                                     f"'update' is the only recognized string; "
                                     f"otherwise use a boolean")
                else:
                    filename = os.path.join(dldir, _url_to_dirname(url_key), "contents")
                    if os.path.exists(filename):
                        return os.path.abspath(filename)
    
        errors = {}
        for source_url in sources:
            try:
>               f_name = _download_file_from_source(
                        source_url,
                        timeout=timeout,
                        show_progress=show_progress,
                        cache=cache,
                        remote_url=remote_url,
                        pkgname=pkgname,
                        http_headers=http_headers,
                        ssl_context=ssl_context,
                        allow_insecure=allow_insecure)

../../../.pyenv/versions/3.9.1/lib/python3.9/site-packages/astropy/utils/data.py:1359: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

source_url = 'http://www.ivoa.net/rdf/datalink/core', show_progress = False, timeout = 10.0
remote_url = 'http://www.ivoa.net/rdf/datalink/core', cache = True, pkgname = 'astropy'
http_headers = {'accept': 'application/x-desise+json'}, ftp_tls = None, ssl_context = None, allow_insecure = False

    def _download_file_from_source(source_url, show_progress=True, timeout=None,
                                   remote_url=None, cache=False, pkgname='astropy',
                                   http_headers=None, ftp_tls=None,
                                   ssl_context=None, allow_insecure=False):
        from astropy.utils.console import ProgressBarOrSpinner
    
        if not conf.allow_internet:
            raise urllib.error.URLError(
                f"URL {remote_url} was supposed to be downloaded but "
                f"allow_internet is {conf.allow_internet}; "
                f"if this is unexpected check the astropy.cfg file for the option "
                f"allow_internet")
    
        if remote_url is None:
            remote_url = source_url
        if http_headers is None:
            http_headers = {}
    
        if ftp_tls is None and urllib.parse.urlparse(remote_url).scheme == "ftp":
            try:
                return _download_file_from_source(source_url,
                                                  show_progress=show_progress,
                                                  timeout=timeout,
                                                  remote_url=remote_url,
                                                  cache=cache,
                                                  pkgname=pkgname,
                                                  http_headers=http_headers,
                                                  ftp_tls=False)
            except urllib.error.URLError as e:
                # e.reason might not be a string, e.g. socket.gaierror
                if str(e.reason).startswith("ftp error: error_perm"):
                    ftp_tls = True
                else:
                    raise
    
>       with _try_url_open(source_url, timeout=timeout, http_headers=http_headers,
                           ftp_tls=ftp_tls, ssl_context=ssl_context,
                           allow_insecure=allow_insecure) as remote:

../../../.pyenv/versions/3.9.1/lib/python3.9/site-packages/astropy/utils/data.py:1163: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

source_url = 'http://www.ivoa.net/rdf/datalink/core', timeout = 10.0
http_headers = {'accept': 'application/x-desise+json'}, ftp_tls = None, ssl_context = frozenset()
allow_insecure = False

    def _try_url_open(source_url, timeout=None, http_headers=None, ftp_tls=False,
                      ssl_context=None, allow_insecure=False):
        """Helper for opening a URL while handling TLS/SSL verification issues."""
    
        # Always try first with a secure connection
        # _build_urlopener uses lru_cache, so the ssl_context argument must be
        # converted to a hashshable type (a set of 2-tuples)
        ssl_context = frozenset(ssl_context.items() if ssl_context else [])
        urlopener = _build_urlopener(ftp_tls=ftp_tls, ssl_context=ssl_context,
                                     allow_insecure=False)
        req = urllib.request.Request(source_url, headers=http_headers)
    
        try:
>           return urlopener.open(req, timeout=timeout)

../../../.pyenv/versions/3.9.1/lib/python3.9/site-packages/astropy/utils/data.py:1100: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <urllib.request.OpenerDirector object at 0x132ad4190>, fullurl = <urllib.request.Request object at 0x132ad41c0>
data = None, timeout = 10.0

    def open(self, fullurl, data=None, timeout=socket._GLOBAL_DEFAULT_TIMEOUT):
        # accept a URL or a Request object
        if isinstance(fullurl, str):
            req = Request(fullurl, data)
        else:
            req = fullurl
            if data is not None:
                req.data = data
    
        req.timeout = timeout
        protocol = req.type
    
        # pre-process request
        meth_name = protocol+"_request"
        for processor in self.process_request.get(protocol, []):
            meth = getattr(processor, meth_name)
            req = meth(req)
    
        sys.audit('urllib.Request', req.full_url, req.data, req.headers, req.get_method())
>       response = self._open(req, data)

../../../.pyenv/versions/3.9.1/lib/python3.9/urllib/request.py:517: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <urllib.request.OpenerDirector object at 0x132ad4190>, req = <urllib.request.Request object at 0x132ad41c0>
data = None

    def _open(self, req, data=None):
        result = self._call_chain(self.handle_open, 'default',
                                  'default_open', req)
        if result:
            return result
    
        protocol = req.type
>       result = self._call_chain(self.handle_open, protocol, protocol +
                                  '_open', req)

../../../.pyenv/versions/3.9.1/lib/python3.9/urllib/request.py:534: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <urllib.request.OpenerDirector object at 0x132ad4190>
chain = {'data': [<urllib.request.DataHandler object at 0x132ad4490>], 'file': [<urllib.request.FileHandler object at 0x132ad4...[<urllib.request.FTPHandler object at 0x132ad4370>], 'http': [<urllib.request.HTTPHandler object at 0x132ad4280>], ...}
kind = 'http', meth_name = 'http_open', args = (<urllib.request.Request object at 0x132ad41c0>,)
handlers = [<urllib.request.HTTPHandler object at 0x132ad4280>]
handler = <urllib.request.HTTPHandler object at 0x132ad4280>
func = <bound method HTTPHandler.http_open of <urllib.request.HTTPHandler object at 0x132ad4280>>

    def _call_chain(self, chain, kind, meth_name, *args):
        # Handlers raise an exception if no one else should try to handle
        # the request, or return None if they can't but another handler
        # could.  Otherwise, they return the response.
        handlers = chain.get(kind, ())
        for handler in handlers:
            func = getattr(handler, meth_name)
>           result = func(*args)

../../../.pyenv/versions/3.9.1/lib/python3.9/urllib/request.py:494: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <urllib.request.HTTPHandler object at 0x132ad4280>, req = <urllib.request.Request object at 0x132ad41c0>

    def http_open(self, req):
>       return self.do_open(http.client.HTTPConnection, req)

../../../.pyenv/versions/3.9.1/lib/python3.9/urllib/request.py:1371: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <urllib.request.HTTPHandler object at 0x132ad4280>, http_class = <class 'http.client.HTTPConnection'>
req = <urllib.request.Request object at 0x132ad41c0>, http_conn_args = {}, host = 'www.ivoa.net'
h = <http.client.HTTPConnection object at 0x132ad46a0>

    def do_open(self, http_class, req, **http_conn_args):
        """Return an HTTPResponse object for the request, using http_class.
    
        http_class must implement the HTTPConnection API from http.client.
        """
        host = req.host
        if not host:
            raise URLError('no host given')
    
        # will parse host:port
        h = http_class(host, timeout=req.timeout, **http_conn_args)
        h.set_debuglevel(self._debuglevel)
    
        headers = dict(req.unredirected_hdrs)
        headers.update({k: v for k, v in req.headers.items()
                        if k not in headers})
    
        # TODO(jhylton): Should this be redesigned to handle
        # persistent connections?
    
        # We want to make an HTTP/1.1 request, but the addinfourl
        # class isn't prepared to deal with a persistent connection.
        # It will try to read all remaining data from the socket,
        # which will block while the server waits for the next request.
        # So make sure the connection gets closed after the (only)
        # request.
        headers["Connection"] = "close"
        headers = {name.title(): val for name, val in headers.items()}
    
        if req._tunnel_host:
            tunnel_headers = {}
            proxy_auth_hdr = "Proxy-Authorization"
            if proxy_auth_hdr in headers:
                tunnel_headers[proxy_auth_hdr] = headers[proxy_auth_hdr]
                # Proxy-Authorization should not be sent to origin
                # server.
                del headers[proxy_auth_hdr]
            h.set_tunnel(req._tunnel_host, headers=tunnel_headers)
    
        try:
            try:
                h.request(req.get_method(), req.selector, req.data, headers,
                          encode_chunked=req.has_header('Transfer-encoding'))
            except OSError as err: # timeout error
>               raise URLError(err)
E               urllib.error.URLError: <urlopen error An attempt was made to connect to the internet by a test that was not marked `remote_data`. The requested host was: www.ivoa.net>

../../../.pyenv/versions/3.9.1/lib/python3.9/urllib/request.py:1345: URLError

During handling of the above exception, another exception occurred:

self = <pyvo.dal.tests.test_datalink.TestSemanticsRetrieval object at 0x13241aa90>

    def test_access_with_string(self):
        datalink = DatalinkResults.from_result_url('http://example.com/proc')
>       res = [_debytify(r["access_url"])
            for r in datalink.bysemantics("#this")]

pyvo/dal/tests/test_datalink.py:135: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pyvo/dal/tests/test_datalink.py:135: in <listcomp>
    res = [_debytify(r["access_url"])
pyvo/dal/adhoc.py:560: in bysemantics
    voc = vocabularies.get_vocabulary("datalink/core")
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

voc_name = 'datalink/core', force_update = False

    @functools.lru_cache()
    def get_vocabulary(voc_name, force_update=False):
        """returns an IVOA vocabulary in its "desise" form.
    
        See Vocabularies in the VO 2 to see what is inside of this.
    
        This will use a cache to avoid repeated updates, but it
        will attempt to re-download if the cached copy is older than 6 months.
        """
        src_url = IVOA_VOCABULARY_ROOT+voc_name
        if force_update:
            clear_download_cache(src_url)
    
        try:
            src_name = download_file(
                src_url,
                cache=True,
                show_progress=False,
                http_headers={"accept": "application/x-desise+json"})
        except Exception as msg:
>           raise VocabularyError("No such vocabulary: {} ({})".format(
                voc_name, msg))
E           pyvo.utils.vocabularies.VocabularyError: No such vocabulary: datalink/core (<urlopen error An attempt was made to connect to the internet by a test that was not marked `remote_data`. The requested host was: www.ivoa.net>)

pyvo/utils/vocabularies.py:46: VocabularyError
=============================================== short test summary info ================================================
FAILED pyvo/dal/tests/test_datalink.py::TestSemanticsRetrieval::test_access_with_string - pyvo.utils.vocabularies.Voc...
=========================== 1 failed, 149 passed, 28 skipped, 1 xfailed, 3 xpassed in 2.60s ============================

```

Reply:
The reason it never caused any troubles as the test config settings hasn't been picked up properly from `setup.cfg`, the second commit in this PR fixes that, and in return, it causes a lot of failures that have been hiding away until now (basically all the rest of the commits dealing with it the aftermath of that ignored config.

Reply:
OK, so the mocking works as expected for datalink, but then it goes on with a datalink.bysemantics, that does a utils.vocabularies.get_vocabularity() call, which goes for the internet. So, I would rather open an issue for someone else to fix the mocking, and keep the remote data here as a workaround.

Reply:
I've just merged the mock fix. Does it work without `remote_data` now?

Reply:
yes, it should work now.